### PR TITLE
Support bootstrap configs in text proto format, and improve errors.

### PR DESCRIPTION
**Pull request by jmillikin-stripe**
This commit allows `.pb_text` configs to be parsed as text, and improves
the parse failure error message to include (1) text or binary mode, and
(2) the message type being parsed.

I was trying to use the new `--bootstrap-path` flag to get away from
JSON so the config could have comments, and got stuck for a while before
realizing the protobuf was being parsed in _binary_ mode.

### CODE REVIEW

**Comment by mattklein123** [comment]
@jmillikin-stripe will defer to @htuch but I thought there was some reason we didn't want to support text PB (vs. JSON/YAML). He can advise.

If we do stick with this we will need a test.

**Comment by htuch** [comment]
Yes, since text PB is not standardized (it's not even consistent across languages in some corner cases), we decided to use YAML as the standardized textual representation of proto. That said, this is a small PR to give us support for this, which might be useful in some test situation, so sure, why not. We won't be officially supporting it though.

**Comment by htuch** [comment]
BTW, text PB is super ugly for the opaque filter configs, take a look at https://gist.github.com/htuch/aa273c9bf01ba9c6b18353ec8799695e#file-listeners_generated-pb. JSON/YAML representations hide this.

File: test/common/protobuf/utility_test.cc [hunk 1]
```diff
@@ -23,4 +23,21 @@ TEST(UtilityTest, LoadProtoFromFile) {
...
+      ->set_address("1.1.1.1");
+
+  std::string bootstrap_text;
+  ASSERT_TRUE(Protobuf::TextFormat::PrintToString(bootstrap, &bootstrap_text));
+  const std::string filename =
+      TestEnvironment::writeStringToFileForTest("proto.pb_text", bootstrap_text);
+
+  envoy::api::v2::Bootstrap proto_from_file;
+  MessageUtil::loadFromFile(filename, proto_from_file);
+  EXPECT_TRUE(TestUtility::protoEqual(bootstrap, proto_from_file));
```
Comment (mattklein123) [inline test/common/protobuf/utility_test.cc]: nit: please also test the exception case with EXPECT_THROW_WITH_MESSAGE by passing a bogus .pb_text file.

**Comment by jmillikin-stripe** [comment]
OK, tests added.

> Yes, since text PB is not standardized (it's not even consistent across languages in some corner cases)

As one datapoint: text_format encoding of proto3 consistency across {C++, Go, Python, Java} is Good Enough™ for what we'll be using it for, and has strong advantages over YAML (simpler semantics/grammar) and JSON (comments). It's also more practical to generate than binary protobufs for environments that can't easily have `protoc` plumbed into them.

**Comment by jmillikin-stripe** [comment]
Dev experience note: Tests won't build on my MacOS machine because it's missing some Linux tools (`realpath`?), and won't build in a `lyft/envoy-build` container for...mysterious reasons:

```
WARNING: /src/bazel/repositories.bzl:24:5: 
WARNING: /src/bazel/repositories.bzl:25:5: Process terminated by signal 15; also encountered an error while attempting to retrieve output
WARNING: /src/bazel/repositories.bzl:26:5: External dep build exited with return code: 256
WARNING: /src/bazel/repositories.bzl:28:9:  External dependency build failed, check above log for errors and ensure all prerequisites at https://github.com/lyft/envoy/blob/master/bazel/README.md#quick-start-bazel-build-for-developers are met.
```

@htuch, I understand a lot of the current Bazel weirdness is due to constraints from fitting the Envoy build into Google's external infra. At some point I'd like to see if it could be simplified so it's easier to build dependencies in a more Bazel-y way.

**Comment by htuch** [comment]
@jmillikin-stripe we're open to improvements to the build, as long as they don't involve creating a parallel Bazel build system for external deps (e.g. our own custom injected BUILD files via `new_git_repository` for every dependency). That was the state we started in earlier this year, and it was hard to maintain and had questionable build products (none of the tests for these dependencies were run with the hacked up BUILD files).

A pure Bazel dependency tree would be amazing, but I think it's a bit of a long term goal, given that dependencies seem to keep growing and very few projects are Bazel native.

That said, if we could move some parts out of the recursive make step to Bazel for dependencies where there is native Bazel support already (e.g. BoringSSL), which you previously expressed interest in, definitely open to that. We should have a better CI setup soon without the Travis resource limits, so that would open up the opportunity to having less done in prebuilts in CI.
